### PR TITLE
Add aditional ip parsing for ipv6 url brackets

### LINKED ISSUE #433: IPv6 parsing race condition
When trying to connect to an ipv6 adress like "64ff:9b00:100::ac10:52f8"

1. connection fails because literal string is passed to library resulting in invalid URL like
> DeviceConnectionError(<InvalidURL http://64ff:9b00:100::ac10:52f8/shelly>)

2. connection fails because ipv6 url brackets like "[64ff:9b00:100::ac10:52f8]" are not passing validation 
> ValueError: '[64ff:9b00:100::ac10:52f8]' does not appear to be an IPv4 or IPv6 address

I'm looking at the library now to add a check for ipv6 adresses and append the brackets when passing to the connection function.

### PR DESCRIPTION
**Pull request by covert8**
Adds aditional check when parsing ip adresses to check for ipv6 adresses and append "[" brackets if so.

Change URLs to be build by aio URL builder automatically adding brackets if applicable 

Fixes #433

### CODE REVIEW

**Comment by thecode** [comment]
Can you explain why would you need to support IPv6? Shelly devices doesn't support IPv6

**Comment by kzyapkov** [comment]
Recent 1.0.x versions do, local only.

**Comment by covert8** [comment]
I'm running an ipv6 only k8s cluster with nat64 (yes i hate myself), this would support this.

The implementation isn't the most elegant i must say. I could look at implementing it in some other way if this wouldn't suffice.

**Comment by covert8** [comment]
I have updated the tr to make use of yarl URL also from aio. 
This makes it a lot more elegant and hands-off the responsibility to the library to add brackets if needed.

**Comment by kzyapkov** [comment]
Just a heads up, to keep feature sets consistent (and free some flash) we will not be enabling IPv6 on Plus devices (and the upcoming version will disable it). It will be available on the Pro line.

**Comment by covert8** [comment]
I am running this change locally and it works. This seems to be the most elegant solution I can come up with.

**Comment by bieniu** [comment]
Use `black` to format the code, otherwise CI will fail.

I do not use IPv6 so I’m not able to test the change. Maybe you can post a screenshot of example.py running for IPv6 device?

Is there also a change required on the HA integration side?

**Comment by covert8** [comment]
Formatting should be in order now.

Here are the debug logs from my HA
˘``
2023-11-12 12:06:04.771 INFO (MainThread) [homeassistant.components.automation.6u] 6u: Executing step device automation                                                    ││ 2023-11-12 12:07:00.255 DEBUG (MainThread) [aioshelly.rpc_device.wsrpc] recv(64ff:9b00:100::ac10:52f8): {'src': 'shellyplus1pm-d4d4da7c9c0c', 'dst': 'aios-140463188692688 ││ ', 'method': 'NotifyStatus', 'params': {'ts': 1699787220.27, 'switch:0': {'id': 0, 'aenergy': {'by_minute': [0.0, 0.0, 0.0], 'minute_ts': 1699787219, 'total': 0.0}}}}     ││ 2023-11-12 12:07:00.255 DEBUG (MainThread) [aioshelly.rpc_device.wsrpc] Notification: NotifyStatus {'ts': 1699787220.27, 'switch:0': {'id': 0, 'aenergy': {'by_minute': [0 ││ .0, 0.0, 0.0], 'minute_ts': 1699787219, 'total': 0.0}}} 
```

No changes are required to the HA integration.

**Comment by covert8** [comment]
Here are the logs from example.py

˘
   python3 example.py -ip 64ff:9b00:100::ac10:52f8 -u [redacted] -p [redacted] --init -deb
DEBUG:aioshelly.rpc_device.wsrpc:Adding device 64ff:9b00:100::ac10:52f8 to WsServer message subscriptions
DEBUG:aioshelly.rpc_device.wsrpc:Trying to connect to device at 64ff:9b00:100::ac10:52f8
INFO:aioshelly.rpc_device.wsrpc:Connected to 64ff:9b00:100::ac10:52f8
DEBUG:aioshelly.rpc_device.wsrpc:send(64ff:9b00:100::ac10:52f8): {'id': 1, 'method': 'Shelly.GetConfig', 'src': 'aios-140663129236944', 'auth': {'realm': 'shellyplus1pm-d4d4da7c9c0c', 'username': 'admin', 'nonce': 1699785922, 'cnonce': 1699787722, 'response': '6b4f74c80d772e8be3a6a524d94ce29abd1ec85cf7efee67a7b774fda7079abb', 'algorithm': 'SHA-256'}}
DEBUG:aioshelly.rpc_device.wsrpc:recv(64ff:9b00:100::ac10:52f8): {'id': 1, 'src': 'shellyplus1pm-d4d4da7c9c0c', 'dst': 'aios-140663129236944', 'result': {'ble': {'enable': False, 'rpc': {'enable': True}, 'observer': {'enable': False}}, 'cloud': {'enable': True, 'server': 'shelly-63-eu.shelly.cloud:6022/jrpc'}, 'input:0': {'id': 0, 'name': None, 'type': 'switch', 'enable': True, 'invert': False, 'factory_reset': True}, 'mqtt': {'enable': False, 'server': None, 'client_id': 'shellyplus1pm-d4d4da7c9c0c', 'user': None, 'topic_prefix': 'shellyplus1pm-d4d4da7c9c0c', 'rpc_ntf': True, 'status_ntf': False, 'use_client_cert': False, 'enable_rpc': True, 'enable_control': True}, 'switch:0': {'id': 0, 'name': None, 'in_mode': 'detached', 'initial_state': 'restore_last', 'auto_on': False, 'auto_on_delay': 60.0, 'auto_off': False, 'auto_off_delay': 60.0, 'power_limit': 4480, 'voltage_limit': 250, 'autorecover_voltage_errors': True, 'current_limit': 16.0}, 'sys': {'device': {'name': None, 'mac': 'D4D4DA7C9C0C', 'fw_id': '20231106-160250/1.0.8-gdba0ee3', 'discoverable': True, 'eco_mode': False, 'addon_type': None}, 'location': {'tz': 'Africa/Algiers', 'lat': 36.7405, 'lon': 3.1159}, 'debug': {'level': 2, 'file_level': None, 'mqtt': {'enable': False}, 'websocket': {'enable': True}, 'udp': {'addr': None}}, 'ui_data': {}, 'rpc_udp': {'dst_addr': None, 'listen_port': None}, 'sntp': {'server': 'time.google.com'}, 'cfg_rev': 18}, 'wifi': {'ap': {'ssid': 'ShellyPlus1PM-D4D4DA7C9C0C', 'is_open': True, 'enable': False, 'range_extender': {'enable': False}}, 'sta': {'ssid': 'iot', 'is_open': False, 'enable': True, 'ipv4mode': 'dhcp', 'ip': None, 'netmask': None, 'gw': None, 'nameserver': None}, 'sta1': {'ssid': None, 'is_open': True, 'enable': False, 'ipv4mode': 'dhcp', 'ip': None, 'netmask': None, 'gw': None, 'nameserver': None}, 'roam': {'rssi_thr': -80, 'interval': 60}}, 'ws': {'enable': False, 'server': None, 'ssl_ca': 'ca.pem'}}}
DEBUG:aioshelly.rpc_device.wsrpc:Shelly.GetConfig(None) -> {'id': 1, 'src': 'shellyplus1pm-d4d4da7c9c0c', 'dst': 'aios-140663129236944', 'result': {'ble': {'enable': False, 'rpc': {'enable': True}, 'observer': {'enable': False}}, 'cloud': {'enable': True, 'server': 'shelly-63-eu.shelly.cloud:6022/jrpc'}, 'input:0': {'id': 0, 'name': None, 'type': 'switch', 'enable': True, 'invert': False, 'factory_reset': True}, 'mqtt': {'enable': False, 'server': None, 'client_id': 'shellyplus1pm-d4d4da7c9c0c', 'user': None, 'topic_prefix': 'shellyplus1pm-d4d4da7c9c0c', 'rpc_ntf': True, 'status_ntf': False, 'use_client_cert': False, 'enable_rpc': True, 'enable_control': True}, 'switch:0': {'id': 0, 'name': None, 'in_mode': 'detached', 'initial_state': 'restore_last', 'auto_on': False, 'auto_on_delay': 60.0, 'auto_off': False, 'auto_off_delay': 60.0, 'power_limit': 4480, 'voltage_limit': 250, 'autorecover_voltage_errors': True, 'current_limit': 16.0}, 'sys': {'device': {'name': None, 'mac': 'D4D4DA7C9C0C', 'fw_id': '20231106-160250/1.0.8-gdba0ee3', 'discoverable': True, 'eco_mode': False, 'addon_type': None}, 'location': {'tz': 'Africa/Algiers', 'lat': 36.7405, 'lon': 3.1159}, 'debug': {'level': 2, 'file_level': None, 'mqtt': {'enable': False}, 'websocket': {'enable': True}, 'udp': {'addr': None}}, 'ui_data': {}, 'rpc_udp': {'dst_addr': None, 'listen_port': None}, 'sntp': {'server': 'time.google.com'}, 'cfg_rev': 18}, 'wifi': {'ap': {'ssid': 'ShellyPlus1PM-D4D4DA7C9C0C', 'is_open': True, 'enable': False, 'range_extender': {'enable': False}}, 'sta': {'ssid': 'iot', 'is_open': False, 'enable': True, 'ipv4mode': 'dhcp', 'ip': None, 'netmask': None, 'gw': None, 'nameserver': None}, 'sta1': {'ssid': None, 'is_open': True, 'enable': False, 'ipv4mode': 'dhcp', 'ip': None, 'netmask': None, 'gw': None, 'nameserver': None}, 'roam': {'rssi_thr': -80, 'interval': 60}}, 'ws': {'enable': False, 'server': None, 'ssl_ca': 'ca.pem'}}}
DEBUG:aioshelly.rpc_device.wsrpc:send(64ff:9b00:100::ac10:52f8): {'id': 2, 'method': 'Shelly.GetStatus', 'src': 'aios-140663129236944', 'dst': 'shellyplus1pm-d4d4da7c9c0c', 'auth': {'realm': 'shellyplus1pm-d4d4da7c9c0c', 'username': 'admin', 'nonce': 1699785922, 'cnonce': 1699787722, 'response': '6b4f74c80d772e8be3a6a524d94ce29abd1ec85cf7efee67a7b774fda7079abb', 'algorithm': 'SHA-256'}}
DEBUG:aioshelly.rpc_device.wsrpc:recv(64ff:9b00:100::ac10:52f8): {'id': 2, 'src': 'shellyplus1pm-d4d4da7c9c0c', 'dst': 'aios-140663129236944', 'result': {'ble': {}, 'cloud': {'connected': True}, 'input:0': {'id': 0, 'state': False}, 'mqtt': {'connected': False}, 'switch:0': {'id': 0, 'source': 'WS_in', 'output': False, 'apower': 0.0, 'voltage': 240.0, 'current': 0.0, 'aenergy': {'total': 0.0, 'by_minute': [0.0, 0.0, 0.0], 'minute_ts': 1699787721}, 'temperature': {'tC': 43.2, 'tF': 109.8}}, 'sys': {'mac': 'D4D4DA7C9C0C', 'restart_required': False, 'time': '12:15', 'unixtime': 1699787722, 'uptime': 234333, 'ram_size': 246004, 'ram_free': 131332, 'fs_size': 458752, 'fs_free': 147456, 'cfg_rev': 18, 'kvs_rev': 0, 'schedule_rev': 0, 'webhook_rev': 0, 'available_updates': {}}, 'wifi': {'sta_ip': '172.16.82.248', 'status': 'got ip', 'ssid': 'iot', 'rssi': -72}, 'ws': {'connected': False}}}
DEBUG:aioshelly.rpc_device.wsrpc:Shelly.GetStatus(None) -> {'id': 2, 'src': 'shellyplus1pm-d4d4da7c9c0c', 'dst': 'aios-140663129236944', 'result': {'ble': {}, 'cloud': {'connected': True}, 'input:0': {'id': 0, 'state': False}, 'mqtt': {'connected': False}, 'switch:0': {'id': 0, 'source': 'WS_in', 'output': False, 'apower': 0.0, 'voltage': 240.0, 'current': 0.0, 'aenergy': {'total': 0.0, 'by_minute': [0.0, 0.0, 0.0], 'minute_ts': 1699787721}, 'temperature': {'tC': 43.2, 'tF': 109.8}}, 'sys': {'mac': 'D4D4DA7C9C0C', 'restart_required': False, 'time': '12:15', 'unixtime': 1699787722, 'uptime': 234333, 'ram_size': 246004, 'ram_free': 131332, 'fs_size': 458752, 'fs_free': 147456, 'cfg_rev': 18, 'kvs_rev': 0, 'schedule_rev': 0, 'webhook_rev': 0, 'available_updates': {}}, 'wifi': {'sta_ip': '172.16.82.248', 'status': 'got ip', 'ssid': 'iot', 'rssi': -72}, 'ws': {'connected': False}}}
** shellyplus1pm-d4d4da7c9c0c - Shelly Plus 1PM  @ 64ff:9b00:100::ac10:52f8 **

Status: {'ble': {}, 'cloud': {'connected': True}, 'input:0': {'id': 0, 'state': False}, 'mqtt': {'connected': False}, 'switch:0': {'id': 0, 'source': 'WS_in', 'output': False, 'apower': 0.0, 'voltage': 240.0, 'current': 0.0, 'aenergy': {'total': 0.0, 'by_minute': [0.0, 0.0, 0.0], 'minute_ts': 1699787721}, 'temperature': {'tC': 43.2, 'tF': 109.8}}, 'sys': {'mac': 'D4D4DA7C9C0C', 'restart_required': False, 'time': '12:15', 'unixtime': 1699787722, 'uptime': 234333, 'ram_size': 246004, 'ram_free': 131332, 'fs_size': 458752, 'fs_free': 147456, 'cfg_rev': 18, 'kvs_rev': 0, 'schedule_rev': 0, 'webhook_rev': 0, 'available_updates': {}}, 'wifi': {'sta_ip': '172.16.82.248', 'status': 'got ip', 'ssid': 'iot', 'rssi': -72}, 'ws': {'connected': False}}
Event: None
Connected: True
`

**Comment by bieniu** [comment]
For Gen1 device I get this exception:

```
$ python3 example.py --ip_address 192.168.2.195 --gen1 --init --username REDACTED --password REDACTED
Traceback (most recent call last):
  File "/home/maciek/develop/aioshelly/example.py", line 319, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/maciek/develop/aioshelly/example.py", line 313, in main
    await test_single(options, args.init, gen)
  File "/home/maciek/develop/aioshelly/example.py", line 62, in test_single
    device = await create_device(aiohttp_session, options, init, gen)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maciek/develop/aioshelly/example.py", line 50, in create_device
    return await BlockDevice.create(aiohttp_session, coap_context, options, init)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maciek/develop/aioshelly/aioshelly/block_device/device.py", line 103, in create
    await instance.initialize()
  File "/home/maciek/develop/aioshelly/aioshelly/block_device/device.py", line 131, in initialize
    await self.update_settings()
  File "/home/maciek/develop/aioshelly/aioshelly/block_device/device.py", line 262, in update_settings
    self._settings = await self.http_request("get", "settings")
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maciek/develop/aioshelly/aioshelly/block_device/device.py", line 289, in http_request
    URL.build(scheme="http", host=self.options.ip_address, path=path),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/maciek/develop/aioshelly/venv/lib/python3.11/site-packages/yarl/_url.py", line 272, in build
    cls._validate_authority_uri_abs_path(host=host, path=path)
  File "/home/maciek/develop/aioshelly/venv/lib/python3.11/site-packages/yarl/_url.py", line 710, in _validate_authority_uri_abs_path
    raise ValueError(
ValueError: Path in a URL with authority should start with a slash ('/') if set
```

File: aioshelly/block_device/device.py [hunk 1]
```diff
@@ -285,7 +286,7 @@ async def http_request(
         try:
             resp: ClientResponse = await self.aiohttp_session.request(
                 method,
-                f"http://{self.options.ip_address}/{path}",
+                URL.build(scheme="http", host=self.options.ip_address, path=path),
```
Comment (bieniu) [inline aioshelly/block_device/device.py:289]:
```suggestion
                URL.build(scheme="http", host=self.options.ip_address, path=f"/{path}"),
```

Comment (covert8) [reply aioshelly/block_device/device.py:289]: You're correct it should have a "/", weird that it didn't trigger any issues in my setup.

**Review by bieniu — APPROVED** [review]
LGTM, tested with gen 1 and 2 devices (only IPv4) and works without problems. Thanks @covert8

**Comment by covert8** [comment]
Thanks for working with me, really beginning to doubt my own decision making skills going ipv6 only.

**Review by chemelli74 — APPROVED** [review]
Thx!

**Pull request merged**